src/main.rs

```diff
@@ -1,42 +1,151 @@
 #![no_std] // from template
 #![no_main] // from template
 
+use core::borrow::{Borrow, BorrowMut};
+
+use panic_probe as _;
+
 use flash_algorithm::*; // from template
 use rtt_target::{rprintln, rtt_init_print}; // from template
 
-mod cmds;
-mod macros;
-
 use stm32h7xx_hal::gpio::Speed;
 use stm32h7xx_hal::pac::QUADSPI;
-// use stm32h7xx_hal::xspi::BankSelect;
-use stm32h7xx_hal::{pac, prelude::*, xspi::Qspi, xspi::QspiMode, xspi::QspiWord};
+use stm32h7xx_hal::xspi::BankSelect;
+use stm32h7xx_hal::{pac, prelude::*, xspi::Qspi, xspi::QspiError, xspi::QspiMode, xspi::QspiWord};
 
 // struct Algorithm; // from template
 
 pub struct Algorithm {
     quadspi: Qspi<QUADSPI>,
 }
 
+
+// MT25QL512ABB
 // from initialization:
 algorithm!(Algorithm, {
+    device_name: "MT25QL512ABB",
+    device_type: DeviceType::ExtSpi,
     flash_address: 0x90000000,
-    flash_size: 0x8000000,
+    flash_size: 0x4000000,
     page_size: 0x100,
     empty_value: 0xFF,
+    program_time_out: 1000,
+    erase_time_out: 20000,
     sectors: [{
-        size: 0x10000,
```

Comment:
sector size is FFFFh, see [data sheet of memory](https://www.micron.com/content/dam/micron/global/secure/products/data-sheet/nor-flash/serial-nor/mt25q/die-rev-b/mt25q-qlkt-l-512-abb-0.pdf) p. 19.
The sector size is exclusive to my research, see https://github.com/probe-rs/webpage/issues/154, which leads to 0x10000.